### PR TITLE
Jdp230102 3 cart controller

### PR DESCRIPTION
**Pull request by andwis93**

### CODE REVIEW

File: src/main/java/com/kodilla/ecommercee/controller/CartController.java [hunk 1]
```diff
@@ -2,52 +2,66 @@
...
+        // Cart cart = findById(cartId)
+        // cart.getProductList.add(findProductById(productId);
+        // above is just rought representation of execution.
+        return new ResponseEntity<>("Product: " + productId + " was added to cart: " + cartId, HttpStatus.OK);
     }
 
-    @DeleteMapping(value = "/deleteItem/{cartItemId}")
-    public ResponseEntity<Void> removeCartItem(@PathVariable Long cartItemId,@RequestParam Long cartId) {
-        return ResponseEntity.ok().build();
+    @PutMapping("/remove{cartId}")
```
Comment (gawrysiakg) [inline src/main/java/com/kodilla/ecommercee/controller/CartController.java]: I'm not sure about @PutMapping but OK, let it be 

 @PutMapping("/remove/{cartId}")    (should add / )

File: src/main/java/com/kodilla/ecommercee/controller/CartController.java [hunk 2]
```diff
@@ -2,52 +2,66 @@
...
+    public ResponseEntity<String> removeProductToCart(@PathVariable long cartId, @RequestParam long productId) {
```
Comment (gawrysiakg) [inline src/main/java/com/kodilla/ecommercee/controller/CartController.java]: removeProductFromCart looks better, but it's only suggestion

**Review by gawrysiakg — APPROVED** [review]
Good job

File: src/main/java/com/kodilla/ecommercee/controller/CartController.java (hunk 1 above)
Comment (Feliscatus8) [reply src/main/java/com/kodilla/ecommercee/controller/CartController.java]: I think it should be @DeleteMapping

Comment (Feliscatus8) [reply src/main/java/com/kodilla/ecommercee/controller/CartController.java]: And you need CartItemId, not CartId - a user can only access their own cart so it's unnecessary to specify which one it is, instead we need to know which product to delete

Comment (viepovsky) [reply src/main/java/com/kodilla/ecommercee/controller/CartController.java]: I am thinking if you're right. In database we need to know which cart is it so we can delete or add product to desired cart. If user can access only his own cart then this method should know which one it is. And how can this controller method know it without giving it cartId? Honestly I have no idea maybe you guys do have?

File: src/main/java/com/kodilla/ecommercee/controller/CartController.java [hunk 3]
```diff
@@ -2,52 +2,66 @@
...
+        cartItemDtos.add(new CartItemDto(
+                2L,"product2","description2", new BigDecimal(200),3L));
+        return ResponseEntity.ok(cartItemDtos);
     }
 
-    @PostMapping(value = "{cartId}")
-    public ResponseEntity<Void> addCartItem(@PathVariable Long cartId, @RequestBody CartItemDto cartItemDto) {
-        //post item into Cart//
-        return ResponseEntity.ok().build();
+    @PutMapping("/add/{cartId}")
```
Comment (viepovsky) [inline src/main/java/com/kodilla/ecommercee/controller/CartController.java:36]: I think that if we don't have to use cartId in method removeProductToCart then here also but for now we must decide if it's ok

File: src/main/java/com/kodilla/ecommercee/controller/CartController.java (hunk 2 above)
Comment (viepovsky) [reply src/main/java/com/kodilla/ecommercee/controller/CartController.java]: I agree

File: src/main/java/com/kodilla/ecommercee/controller/CartController.java [hunk 4]
```diff
@@ -2,52 +2,66 @@
...
-@RequestMapping(consumes = MediaType.APPLICATION_JSON_VALUE, value = "/carts")
+@RequestMapping("/carts")
 public class CartController {
     
-    @PostMapping(value = "/createCart/{userId}")
-    public ResponseEntity<Void> createCart(@PathVariable Long userId) {
-        List<CartItemDto> cart = new ArrayList<>();
-        //post to repository//
-        return ResponseEntity.ok().build();
+    @PostMapping(consumes = MediaType.APPLICATION_JSON_VALUE)
```
Comment (viepovsky) [inline src/main/java/com/kodilla/ecommercee/controller/CartController.java:21]: I am thinking if this method needs getting JSON or we can simplify it by @PathVariable or @RequestParam with userId? What do you think?

File: src/main/java/com/kodilla/ecommercee/controller/CartController.java (hunk 1 above)
Comment (gawrysiakg) [reply src/main/java/com/kodilla/ecommercee/controller/CartController.java]: This endpoint must be simple. Specification says that we should remove product from cart, so we use CartItemId and CartId. We should talk about it together, this endpoint is very problematic.

Comment (MMacina) [reply src/main/java/com/kodilla/ecommercee/controller/CartController.java]: I'm not sure if it is correct but I would do like that:
    @DeleteMapping("/{cartId}/{cartItemId}")
    public ResponseEntity<String> removeItemFromCart(@PathVariable Long cartId, @PathVariable Long cartItemId) {
        return new ResponseEntity<>("Item: " + cartItemId + " was removed from cart: " + cartId, HttpStatus.OK);
    }

Comment (gawrysiakg) [reply src/main/java/com/kodilla/ecommercee/controller/CartController.java]: Maybe you are right

File: src/main/java/com/kodilla/ecommercee/controller/CartController.java (hunk 4 above)
Comment (MMacina) [reply src/main/java/com/kodilla/ecommercee/controller/CartController.java:21]: What would you say for this:
    @PostMapping("/{userId}/{cartId}")
    public ResponseEntity<CartDto> createCart(@PathVariable Long userId, @PathVariable Long cartId) {
        CartDto newCartDto = new CartDto(cartId, userId);
        return ResponseEntity.ok(newCartDto);
    }
:)

File: src/main/java/com/kodilla/ecommercee/controller/CartController.java [hunk 5]
```diff
@@ -2,52 +2,66 @@
...
     }
     
-    @GetMapping(value = "/{userId}/{cartId}")
-    public ResponseEntity<CartDto> getCart(@PathVariable Long userId, @PathVariable Long cartId) {
-        //get from repository//
-        //map to Dto//
-        CartDto currentCard = new CartDto(userId, cartId, new ArrayList<>());
-        return ResponseEntity.ok(currentCard);
+    @GetMapping("/products")
+    public ResponseEntity<List<CartItemDto>> getCart(@RequestParam long cartId) {
```
Comment (MMacina) [inline src/main/java/com/kodilla/ecommercee/controller/CartController.java]: IMHO I will modify this:
    @GetMapping("/cart/{cartId}")
    public ResponseEntity<List<CartItemDto>> getCart(@PathVariable Long cartId) {

File: src/main/java/com/kodilla/ecommercee/controller/CartController.java [hunk 6]
```diff
@@ -2,52 +2,66 @@
...
     }
 
-    @PostMapping(value = "/{cartId}/order")
-    public ResponseEntity<CartItemDto> createOrder(@PathVariable Long cartId) {
-        //Method will return OrderDto when it is ready//
-        //Get current cart from repository//
-        //Map current cart to OrderDto//
-        return ResponseEntity.ok().build();
+    @PostMapping("/createOrder")
+    public ResponseEntity<OrderDto> createOrder(@RequestParam long cartId) {
```
Comment (MMacina) [inline src/main/java/com/kodilla/ecommercee/controller/CartController.java:55]: I would try:
    @PostMapping("/createOrder/{cartId}")
    public ResponseEntity<OrderDto> createOrder(@PathVariable Long cartId) {

**Review by MMacina — COMMENTED** [review]
It seems Cart Controller success will have many fathers :)

**Comment by gawrysiakg** [comment]
> It seems Cart Controller success will have many fathers :)

This whole project will have many fathers, I guess that's a good thing

File: src/main/java/com/kodilla/ecommercee/controller/CartController.java (hunk 6 above)
Comment (andwis93) [reply src/main/java/com/kodilla/ecommercee/controller/CartController.java:55]: With @RequestParam you can have it like this: http://localhost:8080/carts/createOrder?cartId=5. You put your parameter cartId after ?cartID=. I don't know what is better @PathVariable or @RequestParam.

File: src/main/java/com/kodilla/ecommercee/controller/CartController.java (hunk 2 above)
Comment (andwis93) [reply src/main/java/com/kodilla/ecommercee/controller/CartController.java]: Good idea. I copy past above one and didn't change it.

File: src/main/java/com/kodilla/ecommercee/controller/CartController.java (hunk 3 above)
Comment (andwis93) [reply src/main/java/com/kodilla/ecommercee/controller/CartController.java:36]: I'm not sure ether but I think when we will be doing method in java we will need that parameter

File: src/main/java/com/kodilla/ecommercee/controller/CartController.java (hunk 5 above)
Comment (andwis93) [reply src/main/java/com/kodilla/ecommercee/controller/CartController.java]: You might be right. Lets use @PathVariable here.

File: src/main/java/com/kodilla/ecommercee/controller/CartController.java (hunk 4 above)
Comment (andwis93) [reply src/main/java/com/kodilla/ecommercee/controller/CartController.java:21]: I don't know. I'm not sure how it works on frontend side. You have to pass information about user. You probably can do it both ways with JSON and with @PathVariable but I'm not 100%. For now I will just live it like that. 
Marcin: With Post you will not need to pass cartId information. cartId will be generated automatically in Cart class with  @GeneratedValue.

File: src/main/java/com/kodilla/ecommercee/controller/CartController.java (hunk 5 above)
Comment (gawrysiakg) [reply src/main/java/com/kodilla/ecommercee/controller/CartController.java]: Yes, PathVariable will be better

**Review by DawidNst — APPROVED** [review]

**Review by DawidNst — APPROVED** [review]

**Review by MMacina — APPROVED** [review]

**Pull request merged**